<!DOCTYPE html><html><head><meta charset="utf-8"><title>README.md</title><style></style></head><body id="preview">
<hr>
<p>layout: archive
title: “Mr. Eel: Mutation Rate Annotation Utility”
output:
html_document:
toc: true</p>
<hr>
<h3><a id="Overview_8"></a>Overview</h3>
<p><strong>Please note that this web app is in beta, and is presently hosted on an Amazon EC2 micro instance with limited CPU and RAM. As such, problems may arise if multiple sessions are running concurrently. We apologize for any inconvenience.</strong></p>
<p>Mr. Eel is an annotation utility for <strong>M</strong>utation <strong>R</strong>ate <strong>E</strong>stimation using <strong>E</strong>xtremely rare variants and <strong>L</strong>ocal sequence context. It is designed to quickly annotate single-nucleotide variants with their estimated mutation rate, using the method described in <a href="#">citation pending</a>.</p>
<h3><a id="Formatting_Input_13"></a>Formatting Input</h3>
<p>Files must be uploaded as a tab-delimited text file, and <strong>must</strong> start with the following 4 columns:</p>
<table class="table table-striped table-bordered">
<thead>
<tr>
<th>CHR</th>
<th>POS</th>
<th>REF</th>
<th>ALT</th>
<th>-</th>
<th>-</th>
<th>-</th>
</tr>
</thead>
<tbody>
<tr>
<td>-</td>
<td>-</td>
<td>-</td>
<td>-</td>
<td>-</td>
<td>-</td>
<td>-</td>
</tr>
</tbody>
</table>
<ul>
<li>Other columns may be included in the input file, and will be preserved in the output.</li>
<li>Input files can be processed with or without a column header in the first row. If your input file contains a header, the same column names will carry over to the output. If your input file contains additional columns but no header, arbitrary names will be assigned to these columns in the output.</li>
<li>Only autosomal positions will be processed; positions on sex chromosomes, mitochondrial DNA, or unplaced contigs will be discarded. Input data does not need to be sorted, but performance will be <strong>much</strong> faster using sorted input.</li>
<li>An example input file is available <a href="http://www.jedidiahcarlson.com/assets/test_sites.txt">here</a></li>
<li>Sites should be mapped to build 37 (hg19) of the human reference genome. Specifically, this utility uses the 1000 Genomes assembly file, available at <a href="ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/">human_g1k_v37.fasta</a>. Support for other builds will be implemented in a future release.</li>
<li>VCF files are not currently supported, but can be quickly converted to the desired format using <a href="https://samtools.github.io/bcftools/bcftools.html">bcftools</a>:</li>
</ul>
<pre><code>bcftools query -f '%CHROM\t%POS\t%REF\t%ALT\n' in.vcf &gt; positions.txt&quot;
</code></pre>
<h3><a id="Output_31"></a>Output</h3>
<p>The processed output will be identical to the input, with the addition of a column containing the estimated relative mutation rate for each site:</p>
<table class="table table-striped table-bordered">
<thead>
<tr>
<th>CHR</th>
<th>POS</th>
<th>REF</th>
<th>ALT</th>
<th>CATEGORY</th>
<th>MU</th>
<th>-</th>
</tr>
</thead>
<tbody>
<tr>
<td>1</td>
<td>767658</td>
<td>G</td>
<td>T</td>
<td>GC_TA</td>
<td>0.00338503160303719</td>
<td>-</td>
</tr>
<tr>
<td>1</td>
<td>771925</td>
<td>G</td>
<td>A</td>
<td>GC_AT</td>
<td>0.101687967205209</td>
<td>-</td>
</tr>
<tr>
<td>1</td>
<td>871509</td>
<td>G</td>
<td>C</td>
<td>GC_CG</td>
<td>0.00344711848109074</td>
<td>-</td>
</tr>
<tr>
<td>1</td>
<td>904483</td>
<td>G</td>
<td>A</td>
<td>GC_AT</td>
<td>0.00584652048509416</td>
<td>-</td>
</tr>
</tbody>
</table>
<p>If the <strong>Include sequence motif?</strong> option is selected, an additional column containing the sequence motif (and its reverse complement) will also be included:</p>
<table class="table table-striped table-bordered">
<thead>
<tr>
<th>CHR</th>
<th>POS</th>
<th>REF</th>
<th>ALT</th>
<th>CATEGORY</th>
<th>MU</th>
<th>MOTIF</th>
</tr>
</thead>
<tbody>
<tr>
<td>1</td>
<td>767658</td>
<td>G</td>
<td>T</td>
<td>GC_TA</td>
<td>0.00338503160303719</td>
<td>AGCCTTT(AAAGGCT)</td>
</tr>
<tr>
<td>1</td>
<td>771925</td>
<td>G</td>
<td>A</td>
<td>GC_AT</td>
<td>0.101687967205209</td>
<td>AGCCGGT(ACCGGCT)</td>
</tr>
<tr>
<td>1</td>
<td>871509</td>
<td>G</td>
<td>C</td>
<td>GC_CG</td>
<td>0.00344711848109074</td>
<td>GCCCCGG(CCGGGGC)</td>
</tr>
<tr>
<td>1</td>
<td>904483</td>
<td>G</td>
<td>A</td>
<td>GC_AT</td>
<td>0.00584652048509416</td>
<td>GTTCTAC(GTAGAAC)</td>
</tr>
</tbody>
</table>
<p>If a value is entered in the <strong>“Scale rates to:”</strong> field, relative mutation rates will be scaled to a per-site, per-generation mutation rate, calibrated to the value entered (x 10&lt;sup&gt;-8&lt;/sup&gt;). If using this option, we recommend a value between 1.0 and 1.5, corresponding to the genome-wide average mutation rates reported by recent studies. If the default (0) is kept, the raw relative mutation rates will be returned. When using the scaling feature, by default Mr. Eel will return only the base of each rate (e.g., 1.25, not 1.25e-08). To output rates in scientific notation, simply check the <strong>Output in scientific notation?</strong> box.</p>
<h3><a id="Performance_52"></a>Performance</h3>
<p>Mr. Eel takes anywhere from 30 seconds to 15 minutes to run. Since the script reads each chromosome into memory sequentially, speed is primarily determined by the number of unique chromosomes in the input, not the total number of positions (e.g., 10,000 variants on a single chromosome will run faster than 100 variants across all 22 chromosomes). To manage bandwidth, input file size is restricted to 4Mb. If your data exceeds 4Mb, please download the <a href="/cgi/download.php?dir=assets&amp;file=mr_eel.pl">command line annotation script</a>.</p>
<h3><a id="Command_line_usage_55"></a>Command line usage</h3>
<p><strong>The command line script is available to <a href="/cgi/download.php?dir=assets&amp;file=mr_eel.pl">download</a>, along with the <a href="/cgi/download.php?dir=assets&amp;file=ERV_7bp_rates.txt">7-mer rate table</a>; the <code>download_ref.pl</code> script described below is under construction. When using the command line utility, you <em>must</em> specify the <code>--rates</code> and <code>--ref</code> options.</strong>
Usage instructions for the command line utility can be viewed with <code>perl mr_eel.pl --help</code>.
Note that the download <strong>does not</strong> contain a reference genome. A helper script, <code>download_ref.pl</code> is included to download the human_g1k_v37.fasta file used by the web app. If you already have a reference genome on your local machine, you may specify the location with the <code>--ref /path/to/reference.fasta</code> argument.</p>
<h3><a id="Help_60"></a>Help</h3>
<p>Feedback, feature requests, and bug reports are welcome! Please refer to the <a href="https://github.com/carjed/jedidiahcarlson.com/issues">Github repository</a> to review open issues or raise a new one.</p>
<h3><a id="Planned_Features_63"></a>Planned Features</h3>
<p>New features will be implemented in the command line script prior to being integrated into the web app. Planned features include:</p>
<ul>
<li>Support for VCF and BED files</li>
<li>Improved validation and error handling</li>
<li>Optimize memory usage</li>
<li>Support for secure uploads</li>
<li>Additional plots and summary</li>
<li>and more!</li>
</ul>
<p>Please see the <a href="#">development branch</a> (currently pending) of the Mr. Eel repository for the latest experimental release.</p>
<h3><a id="Privacy_Policy_75"></a>Privacy Policy</h3>
<p>We do not require any personally identifiable information in order to use this application. Any data uploaded to our server will be deleted immediately after it is processed. If you have sensitive data, we recommend you download the command line utility, as we cannot guarantee a secure connection to our server at this time.</p>

</body></html>